awx/api/serializers.py

```diff
@@ -4750,39 +4720,30 @@ def create(self, validated_data):
             if through_models:
                 obj_through_model.objects.bulk_create(through_models)
 
-        wfj.status = 'pending'
-        wfj.save()
+        wfj.signal_start()
 
         return WorkflowJobSerializer().to_representation(wfj)
 
     def check_organization_permission(self, attrs, request):
         # validate Organization
         # - If the orgs is not set, set it to the org of the launching user
         # - If the user is part of multiple orgs, throw a validation error saying user is part of multiple orgs, please provide one
-        if not request.user.is_superuser:
-            if 'organization' not in attrs or attrs['organization'] == None or attrs['organization'] == '':
-                if Organization.accessible_pk_qs(request.user, 'read_role').count() == 1:
-                    for tup in Organization.accessible_pk_qs(request.user, 'read_role').all():
-                        attrs['organization'] = Organization.objects.filter(id__in=str(tup[0])).first()
-                elif Organization.accessible_pk_qs(request.user, 'read_role').count() > 1:
-                    raise serializers.ValidationError("User has permission to multiple Organizations, please set one of them in the request")
-                else:
-                    raise serializers.ValidationError("User not part of any organization, please assign an organization to assign to the bulk job")
+        if 'organization' not in attrs or attrs['organization'] == None or attrs['organization'] == '':
+            org = Organization.accessible_objects(request.user, 'member_role').first()
+            if org:
+                attrs['organization'] = org
             else:
-                allowed_orgs = set()
-                requested_org = attrs['organization']
-                if request and not request.user.is_superuser:
-                    [allowed_orgs.add(tup[0]) for tup in Organization.accessible_pk_qs(request.user, 'read_role').all()]
-                    if requested_org.id not in allowed_orgs:
-                        raise ValidationError(_(f"Organization {requested_org.id} not found or you don't have permissions to access it"))
+                raise serializers.ValidationError("User not part of any organization, please assign an organization to assign to the bulk job")
+        else:
+            allowed_orgs = set()
+            requested_org = attrs['organization']
+            if request and not request.user.is_superuser:
+                [allowed_orgs.add(tup[0]) for tup in Organization.accessible_pk_qs(request.user, 'read_role').all()]
+                if requested_org.id not in allowed_orgs:
+                    raise ValidationError(_(f"Organization {requested_org.id} not found or you don't have permissions to access it"))
```

Comment:
I think you are missing something related to super users

I think they should be allowed to not set org
but if they do set an org, it should be one that exists

Reply:
So the intent is that if a superuser doesn't provide an organization, it will be set to null?

My idea here is that system admin is a parent of any organization member_role, so it will just grab one in these cases, to reduce the number of different behaviors here. But for superuser I can see null.

Reply:
yes, null/none

Reply:
I think in one of the demo's we discussed this orgs implementation and it was told that if the user is part of multiple orgs we should not choose one but let user know its part of multi orgs and needs to specify one, which is why we implemented it that way.

Reply:
I am getting following error when trying to send a bulk job api request as a normal user

```
tools_awx_1 | ValueError: Cannot assign "(1,)": "WorkflowJob.organization" must be a "Organization" instance.
tools_awx_1 | 2023-03-06 09:14:21,423 ERROR    [f9cd9648] django.request Internal Server Error: /api/v2/bulk/job_launch/
tools_awx_1 | Traceback (most recent call last):
tools_awx_1 |   File "/var/lib/awx/venv/awx/lib64/python3.9/site-packages/django/core/handlers/exception.py", line 47, in inner
tools_awx_1 |     response = get_response(request)
tools_awx_1 |   File "/var/lib/awx/venv/awx/lib64/python3.9/site-packages/django/core/handlers/base.py", line 181, in _get_response
tools_awx_1 |     response = wrapped_callback(request, *callback_args, **callback_kwargs)
tools_awx_1 |   File "/usr/lib64/python3.9/contextlib.py", line 79, in inner
tools_awx_1 |     return func(*args, **kwds)
tools_awx_1 |   File "/var/lib/awx/venv/awx/lib64/python3.9/site-packages/django/views/decorators/csrf.py", line 54, in wrapped_view
tools_awx_1 |     return view_func(*args, **kwargs)
tools_awx_1 |   File "/var/lib/awx/venv/awx/lib64/python3.9/site-packages/django/views/generic/base.py", line 70, in view
tools_awx_1 |     return self.dispatch(request, *args, **kwargs)
tools_awx_1 |   File "/awx_devel/./awx/api/generics.py", line 333, in dispatch
tools_awx_1 |     return super(APIView, self).dispatch(request, *args, **kwargs)
tools_awx_1 |   File "/var/lib/awx/venv/awx/lib64/python3.9/site-packages/rest_framework/views.py", line 509, in dispatch
tools_awx_1 |     response = self.handle_exception(exc)
tools_awx_1 |   File "/var/lib/awx/venv/awx/lib64/python3.9/site-packages/rest_framework/views.py", line 469, in handle_exception
tools_awx_1 |     self.raise_uncaught_exception(exc)
tools_awx_1 |   File "/var/lib/awx/venv/awx/lib64/python3.9/site-packages/rest_framework/views.py", line 480, in raise_uncaught_exception
tools_awx_1 |     raise exc
tools_awx_1 |   File "/var/lib/awx/venv/awx/lib64/python3.9/site-packages/rest_framework/views.py", line 506, in dispatch
tools_awx_1 |     response = handler(request, *args, **kwargs)
tools_awx_1 |   File "/awx_devel/./awx/api/views/bulk.py", line 50, in post
tools_awx_1 |     result = bulkjob_serializer.create(bulkjob_serializer.validated_data)
tools_awx_1 |   File "/awx_devel/./awx/api/serializers.py", line 4690, in create
tools_awx_1 |     wfj = WorkflowJob.objects.create(**validated_data, is_bulk_job=True, launch_type='manual', created_by=launch_user)
tools_awx_1 |   File "/var/lib/awx/venv/awx/lib64/python3.9/site-packages/django/db/models/manager.py", line 85, in manager_method
tools_awx_1 |     return getattr(self.get_queryset(), name)(*args, **kwargs)
tools_awx_1 |   File "/var/lib/awx/venv/awx/lib64/python3.9/site-packages/django/db/models/query.py", line 451, in create
tools_awx_1 |     obj = self.model(**kwargs)
tools_awx_1 |   File "/var/lib/awx/venv/awx/lib64/python3.9/site-packages/polymorphic/models.py", line 194, in __init__
tools_awx_1 |     super().__init__(*args, **kwargs)
tools_awx_1 |   File "/awx_devel/./awx/main/models/base.py", line 307, in __init__
tools_awx_1 |     r = super(PrimordialModel, self).__init__(*args, **kwargs)
tools_awx_1 |   File "/var/lib/awx/venv/awx/lib64/python3.9/site-packages/django/db/models/base.py", line 485, in __init__
tools_awx_1 |     _setattr(self, field.name, rel_obj)
tools_awx_1 |   File "/var/lib/awx/venv/awx/lib64/python3.9/site-packages/django/db/models/fields/related_descriptors.py", line 215, in __set__
tools_awx_1 |     raise ValueError(
tools_awx_1 | ValueError: Cannot assign "(1,)": "WorkflowJob.organization" must be a "Organization" instance.
```

Reply:
I have reverted _functional_ changes to this place, and prior expectations are expected to hold. Marking conversation as resolved.